rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,203 @@
+
+### Transformer Design
+
+The transformer should be completely syntactic. Initially only compiling any `html` template expressions if `html` was imported from `lit` or `lit-html` directly.
+
+Each individual `TemplateResult` is split into two declarations, a `CompiledTemplate` and `CompiledTemplateResult`. The `CompiledTemplate` is hoisted to the module scope, and contains the prepared HTML and template parts for the original `TemplateResult`. The `CompiledTemplateResult` replaces the `TemplateResult` and binds the static `CompiledTemplate` with the dynamic `values` array.
+
+If a part is used that requires a constructor, the compiler
+will also insert an import such as:
+
+```js
+// ...
+import { _$LH as litHtmlPrivate_1 } from "lit-html/private-ssr-support.js";
+const {
+  AttributePart: _$LH_AttributePart,
+  PropertyPart: _$LH_PropertyPart,
+  BooleanAttributePart: _$LH_BooleanAttributePart,
+  EventPart: _$LH_EventPart,
+} = litHtmlPrivate_1;
+```
+
+This transformer will be exported from `@lit-labs/compiler`, and can be used anywhere TypeScript transformers are accepted, which depends on build setup. Below is an example using [Rollup](https://rollupjs.org/) with [@rollup/plugin-typescript](https://www.npmjs.com/package/@rollup/plugin-typescript):
+
+```js
+// File: rollup.config.js
+import typescript from "@rollup/plugin-typescript";
+import { compileLitTemplates } from "@lit-labs/compiler";
+
+export default {
+  // ...
+  plugins: [
+    typescript({
+      transformers: {
+        before: [compileLitTemplates()],
+      },
+    }),
+    // other rollup plugins
+  ],
+};
+```
+
+#### What templates will not be compiled
+
+The initial version of the compiler will not handle compilation of all templates. In the future, handling of all cases would allow us to completely remove the lit-html prepare phase from runtime.
+
+| Name                                 | Example                                       | Why                                                                                                                                                                                                                                                                           |
+| ------------------------------------ | --------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Re-exported `html` tag               | `import {html} from 'my-rexported-lit';`      | Any templates using this `html` tag will not be compiled, as it isn't imported from `lit` or `lit-html`.                                                                                                                                                                      |
```

Comment:
Can you add a note that we will seek to fix this one. I expect a lot of larger teams will hit this is they vend their own base class or module that re-exports their common directives and decorators, etc.

Since you're using the TypeChecker, this should be possible by resoling the original declaration, no?

Reply:
We are using a TypeChecker, however only local to a single file. This makes us compatible with different build systems which only process files one at a time.

The fix for this will likely be inspired by the HTML minifier API, by allowing the user to pass in a custom template detection function. This function receives the ttl tag identifier (e.g. `html` or `customHtml`), and the moduleSpecifier that it was imported from (e.g. `"lit"`, or `"reExportedLit"`). If the detection function returns `true` we try compile the template.

Do you think I should add these details to the RFC, or leave it open as future work based on user feedback?

Reply:
Just add that we're going to make this better.

Reply:
Done – thank you!